test/integration/tracked_watermark_buffer.cc

```diff
@@ -94,5 +130,62 @@ std::pair<uint32_t, uint32_t> TrackedWatermarkBufferFactory::highWatermarkRange(
   return std::make_pair(min_watermark, max_watermark);
 }
 
+bool TrackedWatermarkBufferFactory::waitUntilTotalBufferedExceeds(
+    uint64_t byte_size, std::chrono::milliseconds timeout) {
+  absl::MutexLock lock(&mutex_);
+  auto predicate = [this, byte_size]() ABSL_SHARED_LOCKS_REQUIRED(mutex_) {
+    mutex_.AssertHeld();
+    return total_buffer_size_ >= byte_size;
+  };
+  return mutex_.AwaitWithTimeout(absl::Condition(&predicate), absl::Milliseconds(timeout.count()));
+}
+
+void TrackedWatermarkBufferFactory::removeDanglingAccounts() {
+  auto accounts_it = account_infos_.begin();
+  while (accounts_it != account_infos_.end()) {
+    auto next = std::next(accounts_it);
+
+    // Remove all "dangling" accounts.
+    if (accounts_it->first.use_count() == 1) {
+      account_infos_.erase(accounts_it);
+    }
+
+    accounts_it = next;
+  }
+}
+
+bool TrackedWatermarkBufferFactory::waitUntilEachAccountChargedAtleast(
+    uint64_t byte_size, uint32_t expected_num_accounts, std::chrono::milliseconds timeout) {
+  absl::MutexLock lock(&mutex_);
+  auto predicate = [this, byte_size,
+                    expected_num_accounts]() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+    mutex_.AssertHeld();
+
+    removeDanglingAccounts();
+    if (account_infos_.size() < expected_num_accounts) {
+      return false;
+    }
+
+    for (auto& acc : account_infos_) {
+      if (static_cast<Buffer::BufferMemoryAccountImpl*>(acc.first.get())->balance() < byte_size) {
```

Comment:
Is this call to balance() happening in the right thread?  Remember that buffer_memory_allocated_ is not atomic.

Reply:
Hmm, I've changed how this is done so it should be running in the worker thread.